### PR TITLE
Feature/execute dispose

### PR DESCRIPTION
**Pull request by joaquinmarti**
Executes the dispose function with a mutationobserver (which triggers when changing story inside a component) and on iframe unload (when navigating mdx files).

Needs to revert changes in the studio to avoid executing the dispose function in the workshop addon.

### CODE REVIEW

**Review by gqio — APPROVED** [review]

File: src/index.ts [hunk 1]
```diff
@@ -5,10 +5,19 @@ export async function renderWith(
...
-  const rendered = await render(require, storyResult, storyType, div);
-  if (!rendered) {
+  const dispose = await render(require, storyResult, storyType, div);
+
+  if (typeof dispose === 'function') {
+    const win = div.ownerDocument.defaultView;
+    win.addEventListener("unload", dispose);
+    
+    const observer = new MutationObserver(dispose);
+    observer.observe(div.parentNode, {childList: true});
```
Comment (gqio) [inline src/index.ts:27]: just thinking you do not need to filter children to avoid unmoutimg the wrong one ?

Comment (joaquinmarti) [reply src/index.ts:27]: I believe the div#root is the only child of body, but I'll filter this as a safety mechanism for the future.

Comment (gqio) [reply src/index.ts:27]: yes today but if I have let say an overlay inside the same div than the story ... everything will be unmounted I think no ?

Comment (joaquinmarti) [reply src/index.ts:27]: Exactly, adding some sibling could cause an ugly bug here. Better filtering it to avoid that.

File: src/index.ts [hunk 2]
```diff
@@ -5,10 +5,24 @@ export async function renderWith(
...
-  const rendered = await render(require, storyResult, storyType, div);
-  if (!rendered) {
+  const dispose = await render(require, storyResult, storyType, div);
+
+  if (typeof dispose === 'function') {
+    const win = div.ownerDocument.defaultView;
+    win.addEventListener("unload", dispose);
+
+    const observer = new MutationObserver((records) => {
+      const deletesRoot = !!records.find(record => Array.from(record.removedNodes).find(node => node.id === "root"));
```
Comment (gqio) [inline src/index.ts]: I was thinking about direct ref comparison directly instead of this magic id

Comment (gqio) [reply src/index.ts]: could we use 'for loop' I think it's more readable, dont you ?

Comment (gqio) [reply src/index.ts]: @joaquinmarti do you think it can be done in this PR ?

**Review by gqio — APPROVED** [review]

File: src/index.ts (hunk 2 above)
Comment (joaquinmarti) [reply src/index.ts]: They're already done in this commit one hour ago @gqio 

https://github.com/divriots/studio-story-renderer/pull/4/commits/4a6ee4278c26109ca71c1c4258f4e7ea0d930a88

Comment (joaquinmarti) [reply src/index.ts]: I believe you're leaving comments in an outdated version.

Comment (gqio) [reply src/index.ts]: Sorry it was pending message 🤦

**Review by muryoh — APPROVED** [review]

**Pull request merged**